core/src/main/java/com/linecorp/armeria/server/throttling/tokenbucket/TokenBucketConfig.java

```diff
@@ -0,0 +1,109 @@
+ */
+
+package com.linecorp.armeria.server.throttling.tokenbucket;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+
+import javax.annotation.Nonnull;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpStatus;
+
+/**
+ * Stores configuration of the Token-Bucket algorithm.
+ */
+public class TokenBucketConfig {
+
+    @Nonnull
+    private final BandwidthLimit[] limits;
+    @Nonnull
+    private final Duration retryAfterTimeout;
+
+    /**
+     * Defines throttling configuration comprised of one or more bandwidth limits in accordance to
+     * token-bucket algorithm.
+     *
+     * <h3>Multiple bandwidths:</h3>
+     * It is possible to specify more than one bandwidth per bucket,
+     * and bucket will handle all bandwidth in strongly atomic way.
+     * Strongly atomic means that token will be consumed from all bandwidth or from nothing,
+     * in other words any token can not be partially consumed.
+     * <br> Example of multiple bandwidth:
+     * <pre>{@code
+     * // Adds bandwidth that restricts to consume
+     * // not often than 1000 tokens per 1 minute and
+     * // not often than 100 tokens per second.
+     * TokenBucketConfig config = TokenBucketConfig.builder()
+     *      .limit(1000L, Duration.ofMinutes(1))
+     *      .limit(100L, Duration.ofSeconds(1))
+     *      .retryAfterTimeout(Duration.ofSeconds(10L))
+     *      .build()
+     * }</pre>
+     *
+     * @param retryAfterTimeout the client timeout to retry the attempt
+     * @param limits one or more bandwidth limits to be used by token-bucket algorithm
+     */
+    TokenBucketConfig(@Nonnull Duration retryAfterTimeout, @Nonnull BandwidthLimit... limits) {
+        requireNonNull(retryAfterTimeout, "retryAfterTimeout");
+        if (retryAfterTimeout.isNegative() || retryAfterTimeout.isZero()) {
+            throw new IllegalArgumentException(
+                    "retryAfterTimeout: " + retryAfterTimeout + " (expected: > 0)");
+        }
+        this.retryAfterTimeout = retryAfterTimeout;
+        this.limits = requireNonNull(limits, "limits");
+    }
+
+    /**
+     * Creates a new {@link TokenBucketConfigBuilder}.
+     */
+    public static TokenBucketConfigBuilder builder() {
+        return new TokenBucketConfigBuilder();
+    }
+
+    /**
+     *  Multiple limits applied to the bucket. This may be empty.
+     * @return An array of {@link BandwidthLimit}
+     */
+    @Nonnull
+    public BandwidthLimit[] limits() {
+        return limits;
+    }
+
+    /**
+     * When the bucket limits reached, the incoming requests will be replied
+     * with {@link HttpStatus#TOO_MANY_REQUESTS} that carries {@link HttpHeaderNames#RETRY_AFTER} with
+     * the value returned by this method.
+     * @return A {@link Duration} after which the client may retry again.
+     */
+    @Nonnull
+    public Duration retryAfterTimeout() {
```

Comment:
agree with something I think said before, which is maybe retryAfter can be compositional

Reply:
Could you please clarify?
[Retry-After](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) is a very important mechanism of throttling which Armeria's ThrottlingService/Strategy have not utilized yet. In fact, it should be consistent with client-side circuit breaker and retry policies. E.g. the client-side circuit breaker should react to this header and open the breaker for the specified period of time automatically. Similarly it could be done for the retry.
It's indeed unrelated to Token-Bucket, but it's required for many use cases I have to deal with.

Reply:
I think in general this issue is more advocating on the premise that everything in bucket4j is relevant here vs mentioning which things you need and how you decide to perform things. for example, retry after time implies you have a sensible value to return. this could be throttle specific or otherwise, and how to calculate that value could be parameterized also. while it may be that the team merge your requests without asking how to simplify, how it is used, how less could be viable, it isnt more than just normal analysis to ask these sort of questions. concretely here.. maybe you can provide some rationale for why it is always relevant a time based back off value? you can choose not to answer of course. someone asked me to look at this issue and I did as a favor to a friend. personally I have no dog in this race.

Reply:
I appreciate your comments. I did not try to undermine your feedback in any way. I just tried to clarify what you meant as "compositional".
Regarding the time-based backoff. I did not say it's absolutely always relevant. The implementation I proposed adds this as an additional layer over base ThrottlingService/Strategy. The user may chose to use this layer or not. Besides, the backoff is optional:
```
        if (retryAfterSeconds != null) {
            return ResponseHeaders.of(status, HttpHeaderNames.RETRY_AFTER, retryAfterSeconds);
        } else {
            return ResponseHeaders.of(status);
        }
```
I also gave you example of Armeria's own CircuitBreaker and Retry implementations where it makes a lot of sense to coordinate better with the server side. If the server does not inform of any particular backoff period and the client relies on its own disconnected configuration than the throttling goals may not be achieved by the end-to-end system.
Now regarding calculating the backoff period dynamically. The API I proposed in RetryThrottlingStrategy allows doing that. The value can change dynamically or even become NULL at some point. Although, my current implementation of TokenBucketConfig does not allow such dynamism. I could use some kind of Supplier for that, but at this point it may just overcomplicate things.